File: backend/src/schemas/analytics.py
from pydantic import BaseModel, ConfigDict, Field


class SpendSummary(BaseModel):
    total_spent: float = Field(...,
                               description="Total amount spent in the given period")
    currency: str = Field(..., min_length=3, max_length=3,
                          description="Primary currency code")
    count: int = Field(...,
                       description="Number of transactions in the given period")
    model_config = ConfigDict(from_attributes=True)


class SpendByMonthItem(BaseModel):
    month: str = Field(...,
                       description="Month in YYYY-MM-DD format (start of month)")
    currency: str = Field(..., min_length=3, max_length=3)
    total: float = Field(...)
    model_config = ConfigDict(from_attributes=True)


class CategorySpendItem(BaseModel):
    category_id: str | None = Field(
        default=None, description="Category id or null for uncategorized")
    category_name: str = Field(...,
                               description="Name of the category or 'Other'")
    currency: str = Field(..., min_length=3, max_length=3)
    total: float = Field(...)
    model_config = ConfigDict(from_attributes=True)

File: StMS/lib/api/analytics.ts
import { apiFetch } from "../api";

export type SpendSummary = {
	total_spent: number;
	currency: string;
	count: number;
};
export type SpendByMonthItem = {
	month: string;
	currency: string;
	total: number;
};
export type CategorySpendItem = {
	category_id: string | null;
	category_name: string;
	currency: string;
	total: number;
};

export async function retrieveSpendingAnalytics(
	date_from: string,
	date_to: string,
	by?: "by-month" | "by-category",
): Promise<SpendSummary | SpendByMonthItem[] | CategorySpendItem[]> {
	const endpoint =
		by === "by-month"
			? "/analytics/spend/by-month"
			: by === "by-category"
				? "/analytics/spend/by-category"
				: "/analytics/spend";

	const url = `${endpoint}?date_from=${date_from}&date_to=${date_to}`;
	const data = await apiFetch<unknown>(url, { method: "GET" });

	if (by === "by-month") return (data as SpendByMonthItem[]) || [];
	if (by === "by-category") return (data as CategorySpendItem[]) || [];

	return (
		(data as SpendSummary) ?? { total_spent: 0, currency: "USD", count: 0 }
	);
}
